toolchain/lower/file_context.cpp

```diff
@@ -190,49 +190,31 @@ auto FileContext::GetOrCreateFunction(SemIR::FunctionId function_id,
   return result;
 }
 
-auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id,
-                                    SemIR::SpecificId specific_id)
-    -> llvm::Function* {
-  const auto& function = sem_ir().functions().Get(function_id);
-
-  // Don't lower generic functions. Note that associated functions in interfaces
-  // have `Self` in scope, so are implicitly generic functions.
-  if (function.generic_id.has_value() && !specific_id.has_value()) {
-    return nullptr;
-  }
-
-  // Don't lower builtins.
-  if (function.builtin_function_kind != SemIR::BuiltinFunctionKind::None) {
-    return nullptr;
+auto FileContext::BuildFunctionTypeInfo(
+    const SemIR::Function& function, SemIR::SpecificId specific_id,
+    llvm::ArrayRef<SemIR::InstId> implicit_param_patterns,
+    llvm::ArrayRef<SemIR::InstId> param_patterns,
+    const SemIR::ReturnTypeInfo& return_info,
```

Comment:
What do you think of grouping these 4 out-parameters into a struct and returning it instead of taking a bunch of reference parameters?

Reply:
I figured for one call it wasn't worth it - but perhaps there's some way to make this a bit less awkward... 

I did manage to sink the actual creation of the `llvm::FunctionType*` into the helper function - removing one of the out parameters and have the `FunctionType*` as the return type. It still leaves the three other out parameters (parameter inst ids, return type, return inst id).

Maybe this feels better? With a main result, and some side results?

Open to adding a struct if that seems like the better option to you.

Reply:
Yeah mixing returns and out params, I would prefer the struct, myself